soc/riscv/riscv-privilege/virt/Kconfig.soc

```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2020 Cobham Gaisler AB
+# SPDX-License-Identifier: Apache-2.0
+
+choice
+	prompt "QEMU RISC-V VirtIO Board"
+	depends on SOC_SERIES_RISCV_VIRT
+
+config SOC_RISCV_VIRT
+	bool "QEMU RISC-V VirtIO Board"
+	select ATOMIC_OPERATIONS_BUILTIN
+
+endchoice
```

Comment:
In my understanding, ATOMIC_OPERATIONS_BUILTIN depends on compiler's built-in atomic operation that is using "A" extension of RISC-V. If there is a very simple processor that does not have A extension (ex. rv32i), Zephyr cannot build for it.

Fortunately, we can use C version atomic mechanism if SMP is disabled, so how about this?
        select ATOMIC_OPERATIONS_C if !SMP
        select ATOMIC_OPERATIONS_BUILTIN if SMP

Reply:
From the point of view of the `virt` Zephyr SOC description, using
`ATOMIC_OPERATIONS_BUILTIN` is appropriate because the emulated
hardware has this capability. Only the RISC-V `virt` SOC is affected
by this `Kconfig` `select`.

That being said, you are highlighting a very important topic regarding
available RISC-V extensions and how they can be described and selected
in Zephyr. Currently the `ATOMIC_OPERATIONS_*` settings for RISC-V
appear to be forced on `soc` level.

My take on this is that it should be solved on a higher level in the
Zephyr RISC-V infrastructure, and not being decided by the SOC Kconfig
alone. One approach would be to introducing a capability logic similar
to `CPU_HAS_FPU` (capability) and `CPU` (actual):

* A capability setting for the SOC to define: `CPU_HAS_EXTENSION_A`.
* A configuration setting for the application to define: `APP_WANTS_TO_USE_EXTENSION_A`,
  which is available only if the SOC has set `CPU_HAS_EXTENSION_A`.

It would allow to remove all references to `ATOMIC_OPERATIONS_*`
currently in `soc/riscv` and replace with a single selection on
`arch/riscv` level:

```
select ATOMIC_OPERATIONS_C if !APP_WANTS_TO_USE_EXTENSION_A
select ATOMIC_OPERATIONS_BUILTIN if (SMP || APP_WANTS_TO_USE_EXTENSION_A)
```
Similarly, the new config can be used to construct the ISA string to the
compiler `-march=` option.

Then it will be a straight forward operation to run the test suite with
and without the "A" extension on the virt SOC:

```
twister -v --platform qemu_riscv32 -x=CONFIG_APP_WANTS_TO_USE_EXTENSION_A=n 2> log.a_no
twister -v --platform qemu_riscv32 -x=CONFIG_APP_WANTS_TO_USE_EXTENSION_A=y 2> log.a_yes
```

Similar (capability+actual) could be introduced for the "M" and "C" extensions.

However, I don't see that the above will become part of this
PR. Maybe this could be an enhancement ticket? What do you think?

Reply:
I understand your nice idea. Let's keep this patch current version on this PR and will solve on higher level by other enhancement ticket.